packages/react/src/components/TreeView/TreeNode.js

```diff
@@ -0,0 +1,276 @@
+    [`${prefix}--tree-node--disabled`]: disabled,
+    [`${prefix}--tree-node--selected`]: isSelected,
+    [`${prefix}--tree-node--with-icon`]: Icon,
+    [`${prefix}--tree-leaf-node`]: !children,
+    [`${prefix}--tree-parent-node`]: children,
+  });
+  const toggleClasses = classNames(`${prefix}--tree-parent-node__toggle-icon`, {
+    [`${prefix}--tree-parent-node__toggle-icon--expanded`]: expanded,
+  });
+  const handleToggleClick = (event) => {
+    if (onToggle) {
+      onToggle(event, { id, isExpanded: !expanded, label, value });
+    }
+    setExpanded(!expanded);
+  };
+  const handleClick = (event) => {
+    event.stopPropagation();
+    if (!disabled) {
+      if (onTreeSelect) {
+        onTreeSelect(event, { id, label, value });
+      }
+      if (onNodeSelect) {
+        onNodeSelect(event, { id, label, value });
+      }
+      if (rest.onClick) {
+        rest.onClick(event);
+      }
+    }
+  };
+  const handleKeyDown = (event) => {
+    if (matches(event, [keys.ArrowLeft, keys.ArrowRight, keys.Enter])) {
+      event.stopPropagation();
+    }
+    if (match(event, keys.ArrowLeft)) {
+      const findParentTreeNode = (node) => {
+        if (node.classList.contains(`${prefix}--tree-parent-node`)) {
+          return node;
+        }
+        if (node.classList.contains(`${prefix}--tree`)) {
+          return null;
+        }
+        return findParentTreeNode(node.parentNode);
+      };
+      if (children && expanded) {
+        onToggle(event, { id, isExpanded: false, label, value });
+        setExpanded(false);
+      } else {
+        /**
+         * When focus is on a leaf node or a closed parent node, move focus to
+         * its parent node (unless its depth is level 1)
+         */
+        findParentTreeNode(currentNode.current.parentNode)?.focus();
+      }
+    }
+    if (children && match(event, keys.ArrowRight)) {
+      if (expanded) {
+        /**
+         * When focus is on an expanded parent node, move focus to the first
+         * child node
+         */
+        currentNode.current.lastChild.firstChild.focus();
+      } else {
+        onToggle(event, { id, isExpanded: true, label, value });
+        setExpanded(true);
+      }
+    }
+    if (matches(event, [keys.Enter, keys.Space])) {
+      event.preventDefault();
+      handleClick(event);
+    }
+    if (rest.onKeyDown) {
+      rest.onKeyDown(event);
+    }
+  };
+
+  useEffect(() => {
+    /**
+     * Negative margin shifts node to align with the left side boundary of the
+     * tree
+     * Dynamically calculate padding to recreate tree node indentation
+     * - parent nodes have (depth + 1rem) left padding
+     * - leaf nodes have (depth + 2.5rem) left padding without icons (because
+     *   of expando icon + spacing)
+     * - leaf nodes have (depth + 2rem) left padding with icons (because of
+     *   reduced spacing between the expando icon and the node icon + label)
+     */
+    const calcOffset = () => {
```

Comment:
Just curious, could this layout be achieved in CSS or is there something about it that requires JS?

Reply:
I don't believe this is achievable through CSS only

Reply:
@emyarod what about it prevents it from being in CSS?

Reply:
I think being unable to access the nesting depth in CSS is the main reason